### PR TITLE
Add MosaicML as a provider to MLflow AI Gateway

### PR DESCRIPTION
**Pull request by arpitjasa-db**
## What changes are proposed in this pull request?

Add MosaicML as a provider to MLflow AI Gateway

## How is this patch tested?

- [X] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [x] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add MosaicML as a provider to MLflow AI Gateway

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
Don't forget to add the new provider to the ``config_types`` dict so that the key can be read (it's optional, but only for users) 

https://github.com/mlflow/mlflow/blob/6de6a63cb3edeb519dfbd7fd4fb49ec7d15ac911/mlflow/gateway/config.py#L144-L149

**Comment by BenWilson2** [comment]
Also the ``Model`` constructor will need to update the config allowance type here too: https://github.com/mlflow/mlflow/blob/6de6a63cb3edeb519dfbd7fd4fb49ec7d15ac911/mlflow/gateway/config.py#L194-L204

File: mlflow/gateway/providers/mosaicml.py
Comment (dbczumar) [inline mlflow/gateway/providers/mosaicml.py:23]: Should we make this configurable? Seems like there's a good chance this could change in the medium term :). Using `"https://models.hosted-on.mosaicml.hosting"` as a default makes sense, but users should be able to override

Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:23]: How would we want users to be able to override this? An additional parameter in `create_route` or in the payload when `querying` each time?

Comment (BenWilson2) [reply mlflow/gateway/providers/mosaicml.py:23]: The config can be updated with an override entry for processing from the config.yaml :) I'll add this to the commit I'm going to make in a few minutes.... one sec :)

File: mlflow/gateway/providers/utils.py
Comment (BenWilson2) [inline mlflow/gateway/providers/utils.py:28]: Invalid POST content to MosaicML does not return `application/json` format; instead it returns `text/plain` for reporting the underlying issue.

**Comment by BenWilson2** [comment]
MosaicML completions endpoint (Llama2-70B) validation test:

Screenshot 2023-08-25 at 8 42 34 PM

MosaicML embeddings endpoint validation test:
Screenshot 2023-08-25 at 8 42 53 PM

File: examples/gateway/mosaicml/config.yaml
Comment (vladimirk-db) [inline examples/gateway/mosaicml/config.yaml:1]: Can we add an example w/ llama2-70b-chat?

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:30]: Why not add this now? We are doing a fast follow-up to add chat completions support for the databricks version as well. We should get this in time for mlflow 2.7.0.

(If you don't want to increase the scope of this PR, we can do it in a follow up.)

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:1]: Can you please (manually) the scenario described below works as intended?

https://databricks.atlassian.net/browse/MLOPS-167

That is, we are returning an informative error message as opposed to `"too many 500 error responses"`

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:30]: See https://databricks.atlassian.net/browse/MLOPS-163

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:1]: Note: they will start returning 400s from mosaic on this soon, but let's just make sure the current path works.
https://mosaicml.atlassian.net/browse/MCLOUD-2970

**Review by vladimirk-db — COMMENTED** [review]
Thanks for quickly iterating on this @arpitjasa-db 

Main request on my side is to add mappings for chat completions
(https://databricks.atlassian.net/browse/MLOPS-163) so the UX of prompting llama2-70b-chat is simpler.

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:30]: Once we align on https://github.com/databricks/universe/pull/374719, we can add that support here as well

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:1]: I think this should be a fast follow to not block this for the 2.7.0 release

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: Actually just ended up adding equivalent support here in the interest of time, and will update both as necessary PTAL

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:41]:
```suggestion
   def _parseChatMessagesToPrompt(self, messages: List[chat.RequestMessage]) -> str:
```

Do we need `async`?

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:39]: Can we move this under `_parseChatMessagesToPrompt` like this?

```
def _parseChatMessagesToPrompt():
    """
    This parser ...
```

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:31]: /ai change this to snake case

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:178]: Does `MosaicML` provide toke usage info in the response? If it doesn't, it's fine.

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:45]: Is `.role` missing?

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:49]: Is `$` necessary?

```python
>>> a = 1
>>> f"${a}"
'$1'
```

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:61]: https://github.com/mlflow/mlflow/pull/9459/files#r1308142749 same here

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:51]: What if there's no system role provided? We still would want to wrap the prompt with 
`<INST> .. </INST>` in that case .

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [reply mlflow/gateway/providers/mosaicml.py:45]: /ai any other approach to achieve the same thing?

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:47]: We could possibly simplify this by breaking it up.

Add [INST] starter tag.

If there is system or assistant role messags, add the <SYS> wrapper around them. (What do we do if theres' both or multiple?)

Add the user message (hopefully it's present, maybe validation needed :-))  and close off w/ [INST]

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [reply mlflow/gateway/providers/mosaicml.py:45]: /ai any advantage over the original approach?

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:47]: Or you could do this in the loop, but make sure to pull up adding [INST] as a starter tag outside the loop. We should still clarify what to do w/ duplicated role entries (e.g., multiple system messages -> just concatenate them?).

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [reply mlflow/gateway/providers/mosaicml.py:45]: @mlflow Got it, thanks for the explanation!

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:109]: I think we need to map the full response back as well?
Isn't the model going to return something like this:

```
 "<s>[INST] <<SYS>>
            {{ system_prompt }}
            <</SYS>>
            {{ user_msg_1 }} [/INST] {{ model_answer_1 }} </s>"
```

W/ a `model_answer_1` added to it? Assistant response would be the `model_answer_1`, but we should return the previous `system_prompt` and `user_msg_1` as well?  This way, from the user perspective, we get the full history of messages back and forth.

We can test this out against staging to verify the behavior.

File: mlflow/gateway/providers/mosaicml.py
Comment (harupy) [inline mlflow/gateway/providers/mosaicml.py:40]: Do other MosaicML models use this template?

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:47]: Also, I think <s> </s> separates the entries in a multi turn chat response.

**Review by vladimirk-db — COMMENTED** [review]
Rounding into shape, I added some chat related comments.

**Review by harupy — APPROVED** [review]
Left some comments, otherwise LGTM!

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:109]: From: 
https://huggingface.co/blog/llama2#how-to-prompt-llama-2

"The model is stateless and does not "remember" previous fragments of the conversation, we must always supply it with all the context so the conversation can continue. This is the reason why context length is a very important parameter to maximize, as it allows for longer conversations and larger amounts of information to be used."

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: Idk about AI but I gotchu 😂

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:178]: It does not 😔

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: Good catch! Added

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: Good catch! Removed

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:40]: No just Llama2

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: That's why we add the `INST` if there was no system role provided

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:47]: I followed the format described here so we should be good: https://huggingface.co/blog/llama2#how-to-prompt-llama-2

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py:109]: Nope, it just returns output. Tested it out here: https://huggingface.co/blog/llama2#how-to-prompt-llama-2

File: examples/gateway/mosaicml/README.md
Comment (harupy) [inline examples/gateway/mosaicml/README.md:5]:
```suggestion
This configuration file specifies three routes: 'completions' and 'embeddings', both using MosaicML's models 'mpt-7b-instruct' and 'instructor-xl', respectively.
```

We have three routes.

**Review by BenWilson2 — APPROVED** [review]
LGTM!

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [inline mlflow/gateway/providers/mosaicml.py:53]: Update this,  please:  it should be  `if not prompt.endswith("<</SYS>>"):`

Can we manually QA this before merging? 

Make sure the cases of user instructions with or without system prompt work fine. Thanks!!

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:119]: Ah, ok, I see. Sounds like we functionality for gateway in the future :-)

File: mlflow/gateway/providers/mosaicml.py
Comment (vladimirk-db) [reply mlflow/gateway/providers/mosaicml.py:53]: (If feasible, adding a unit test for these cases is much better!)

File: tests/gateway/providers/test_mosaicml.py
Comment (vladimirk-db) [inline tests/gateway/providers/test_mosaicml.py:239]: Let's add a test case here for no system role provided.

Comment (vladimirk-db) [reply tests/gateway/providers/test_mosaicml.py:239]: (Also no assistant role as well.)
IMHO let's cover at least these:
1. system, user - valid (frequent path)
2. user - valid (frequent path)
4. system by itself (is it valid?)
5. (system, assistant) and (user, assistant)  (are these valid?)

**Review by vladimirk-db — APPROVED** [review]
Approving to unblock - please address all comments before merging though (one is a bug, the others are about adding tests and doing some QA).

Thanks @arpitjasa-db !!

File: mlflow/gateway/providers/mosaicml.py
Comment (arpitjasa-db) [reply mlflow/gateway/providers/mosaicml.py]: Added the case!

File: tests/gateway/providers/test_mosaicml.py
Comment (arpitjasa-db) [reply tests/gateway/providers/test_mosaicml.py]: Added all the cases, except for "system by itself" and "(system, assistant)" which are not valid cases. So added a check for those and cases to cover that

File: tests/gateway/providers/test_mosaicml.py
Comment (vladimirk-db) [inline tests/gateway/providers/test_mosaicml.py:246]: We need closing `</s>` here and in the other tests, right @arpitjasa-db ?

Comment (arpitjasa-db) [reply tests/gateway/providers/test_mosaicml.py:246]: No actually, we're supposed to leave it open for the final instruction
Screenshot 2023-08-29 at 3 08 48 PM
Screenshot 2023-08-29 at 3 08 54 PM

Comment (vladimirk-db) [reply tests/gateway/providers/test_mosaicml.py:246]: Got it!

**Pull request merged**
